feat: retry visibility (#933)
goodeggs/json-fetch: README.md

You can use any attribute of the [FetchResponse](https://developer.mozilla.org/en-US/docs/Web/API/Response) or Error to determine whether to retry or not. Your function _must_ handle both errors (such as network errors) and FetchResponse objects without blowing up. We recommend stateless, side-effect free functions. You do not need to worry about the maximum number of retries -- promise-retry will stop retrying after the maximum you specify. See the tests and `src/retriers.js` file for examples.

### On Request callbacks

We've implemented two callbacks functions that can be passed as part of the **jsonFetch** options to do something `onRequestStart` and `onRequestEnd`, this is specially helpful to log request and response data in retry calls.

All `jsonFetch` options are availables parameters for this two callback functions plus

> ```

Reviewer: ```suggestion
```typescript
```

url?: string;
retryCount?: number;
responseOrError?: Response | Error | unknown;
```

> An usage example of this callbacks fucntions is as follows:

Reviewer: A few typos here
```suggestion
For example, to log before and after each request:
```

> ```js
> let requestUrl = 'http://www.test.com/products/1234'
> await jsonFetch(requestUrl, {
>   onRequestStart: ({url: requestUrl, retryCount: retryCount}) => { console.log(`-> AttemptStart ${retryCount}: ${url} with defined timeout ${timeout}`); },
>   onRequestEnd: ({responseOrError retryCount: retryCount, }) => { console.log(`-> AttemptEnd ${retryCount}: Response ${responseOrError.statusText}`); }
> });
> ```

Reviewer: a few typos and nitpicks:
```suggestion
```typescript
const requestUrl = 'http://www.test.com/products/1234';
await jsonFetch(requestUrl, {
  onRequestStart: ({url, retryCount}) =>
    console.log(`Requesting ${url} with timeout ${timeout}, attempt ${retryCount}`),
  onRequestEnd: ({url, retryCount, responseOrError}) =>
    console.log(`Requested ${url}, attempt ${retryCount}, got status ${responseOrError.status}`),
});
```
```

You can use any attribute of the [FetchResponse](https://developer.mozilla.org/en-US/docs/Web/API/Response) or Error to determine whether to retry or not. Your function _must_ handle both errors (such as network errors) and FetchResponse objects without blowing up. We recommend stateless, side-effect free functions. You do not need to worry about the maximum number of retries -- promise-retry will stop retrying after the maximum you specify. See the tests and `src/retriers.js` file for examples.

### On Request callbacks

> We've implemented two callbacks functions that can be passed as part of the **jsonFetch** options to do something `onRequestStart` and `onRequestEnd`, this is specially helpful to log request and response data in retry calls.
> 
> All `jsonFetch` options are availables parameters for this two callback functions plus
> 
> ```typescript
> url?: string;
> retryCount?: number;
> responseOrError?: Response | Error | unknown;
> ```

Reviewer: Clarifying a bit:
```suggestion
Two callback functions can be passed as options to do something `onRequestStart` and `onRequestEnd`. This may be especially helpful to log request and response data on each request.

If you have retries enabled, these will trigger before and after each _actual, individual request_.

#### `onRequestStart`

If given, `onRequestStart` is called with:
```typescript
{
  // ... all the original json-fetch options, plus:
  url: string;
  retryCount: number;
}
```

#### `onRequestEnd`

If given, `onRequestEnd` is called with:
```typescript
{
  // ... all the original json-fetch options, plus:
  ur?: string;
  retryCount: number;
  responseOrError: Response | Error;
}
```
```

Reviewer: Ugh, GitHub can't handle the syntax of Markdown suggestions because it's Markdown within Markdown... Hopefully you can figure out what I meant!

Author: LOL

For example, to log before and after each request:

```typescript
const requestUrl = 'http://www.test.com/products/1234';
await jsonFetch(requestUrl, {

>   onRequestStart: ({url, retryCount}) =>
>     console.log(`Requesting ${url} with timeout ${timeout}, attempt ${retryCount}`),

Reviewer: ```suggestion
  onRequestStart: ({url, timeout, retryCount}) =>
    console.log(`Requesting ${url} with timeout ${timeout}, attempt ${retryCount}`),
```

  onRequestEnd: ({url, retryCount, responseOrError}) =>
    console.log(`Requested ${url}, attempt ${retryCount}, got status ${responseOrError.status}`),

> });

Reviewer: There's a missing closing "```" here

You can use any attribute of the [FetchResponse](https://developer.mozilla.org/en-US/docs/Web/API/Response) or Error to determine whether to retry or not. Your function _must_ handle both errors (such as network errors) and FetchResponse objects without blowing up. We recommend stateless, side-effect free functions. You do not need to worry about the maximum number of retries -- promise-retry will stop retrying after the maximum you specify. See the tests and `src/retriers.js` file for examples.

### On Request callbacks

Two callback functions can be passed as options to do something `onRequestStart` and `onRequestEnd`. This may be especially helpful to log request and response data on each request.
If you have retries enabled, these will trigger before and after each _actual, individual request_.

#### `onRequestStart`

If given, `onRequestStart` is called with:

```typescript
{
  // ... all the original json-fetch options, plus:
  url: string;
  retryCount: number;
}
```

#### `onRequestEnd`

If given, `onRequestEnd` is called with:

```typescript
{
  // ... all the original json-fetch options, plus:

>   ur?: string;

Reviewer: ```suggestion
  url: string;
```
